backend_new/test/unit/services/listing.service.spec.ts

```diff
@@ -0,0 +1,1972 @@
+import { ListingFilterParams } from '../../../src/dtos/listings/listings-filter-params.dto';
+import { LanguagesEnum } from '@prisma/client';
+import { Unit } from '../../../src/dtos/units/unit-get.dto';
+import { UnitTypeSort } from '../../../src/utilities/unit-utilities';
+
+/*
+  generates a super simple mock listing for us to test logic with
+*/
+const mockListing = (
+  pos: number,
+  genUnits?: { numberToMake: number; date: Date },
+) => {
+  const toReturn = { id: pos, name: `listing ${pos + 1}`, units: undefined };
+  if (genUnits) {
+    const units: Unit[] = [];
+    const { numberToMake, date } = genUnits;
+    for (let i = 0; i < numberToMake; i++) {
+      units.push({
+        id: `unit ${i}`,
+        createdAt: date,
+        updatedAt: date,
+        amiPercentage: `${i}`,
+        annualIncomeMin: `${i}`,
+        monthlyIncomeMin: `${i}`,
+        floor: i,
+        annualIncomeMax: `${i}`,
+        maxOccupancy: i,
+        minOccupancy: i,
+        monthlyRent: `${i}`,
+        numBathrooms: i,
+        numBedrooms: i,
+        number: `unit ${i}`,
+        sqFeet: `${i}`,
+        monthlyRentAsPercentOfIncome: `${i % UnitTypeSort.length}`,
+        bmrProgramChart: !(i % 2),
+        unitTypes: {
+          id: `unitType ${i}`,
+          createdAt: date,
+          updatedAt: date,
+          name: UnitTypeSort[i % UnitTypeSort.length],
+          numBedrooms: i,
+        },
+        unitAmiChartOverrides: {
+          id: `unitAmiChartOverrides ${i}`,
+          createdAt: date,
+          updatedAt: date,
+          items: [
+            {
+              percentOfAmi: i,
+              householdSize: i,
+              income: i,
+            },
+          ],
+        },
+        amiChart: {
+          id: `AMI${i}`,
+          items: [],
+          name: `AMI Name ${i}`,
+          createdAt: date,
+          updatedAt: date,
+        },
+      });
+    }
+    toReturn.units = units;
+  }
+
+  return toReturn;
+};
+
+const mockListingSet = (
+  pos: number,
+  genUnits?: { numberToMake: number; date: Date },
+) => {
+  const toReturn = [];
+  for (let i = 0; i < pos; i++) {
+    toReturn.push(mockListing(i, genUnits));
+  }
+  return toReturn;
+};
+
+describe('Testing listing service', () => {
+  let service: ListingService;
+  let prisma: PrismaService;
+
+  beforeEach(async () => {
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [ListingService, PrismaService],
+    }).compile();
+
+    service = module.get<ListingService>(ListingService);
+    prisma = module.get<PrismaService>(PrismaService);
+  });
+
+  it('testing list() with no params', async () => {
+    prisma.listings.findMany = jest.fn().mockResolvedValue(mockListingSet(10));
+
+    prisma.listings.count = jest.fn().mockResolvedValue(10);
+
+    const params: ListingsQueryParams = {};
+
+    expect(await service.list(params)).toEqual({
+      items: [
+        { id: '0', name: 'listing 1' },
```

Comment:
can this be mocklistingset?

Reply:
Hey @emilyjablonski excellent question!

the service.list() ends up mapping the output of the prisma call to the DTO specified. Because of that, in most cases the out of the prisma call won't match up 1 to 1 with the actual function response (transformations/additional fields may find there way in)


in this case if you notice the mock has a real number for the id, and the response has that mapped to a string. Its not a case we really expect the real backend to go through but this verifies the mapping is happening and the typing for the DTO holds true